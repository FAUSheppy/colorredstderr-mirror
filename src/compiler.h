/*
 * Compiler specific macros.
 *
 * Copyright (C) 2013-2014  Simon Ruderich
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef COMPILER_H
#define COMPILER_H 1

#ifdef HAVE___ATTRIBUTE__
/* Prevent/force inlining. Used to improve performance. */
# define noinline      __attribute__((noinline))
# define always_inline __attribute__((always_inline))
/* Unused parameter. */
# define unused        __attribute__((unused))
/* Mark the function protected, which means it can't be overwritten by other
 * modules (libraries), e.g. with LD_PRELOAD); otherwise same as the default
 * visibility. This causes the compiler not use the PLT (and no relocations)
 * for local calls from inside this module; but the symbols are still
 * exported. This is faster and in this case prevents useless lookups as we
 * hook those functions and nobody else should modify them. Not strictly
 * necessary, but nice to have. */
# define visibility_protected __attribute__((visibility("protected")))
#else
# define noinline
# define always_inline
# define unused
# define visibility_protected
#endif

/* Branch prediction information for the compiler. */
#ifdef HAVE___BUILTIN_EXPECT
# define likely(x)   __builtin_expect(!!(x), 1)
# define unlikely(x) __builtin_expect(!!(x), 0)
#else
# define likely(x)   x
# define unlikely(x) x
#endif

#endif
