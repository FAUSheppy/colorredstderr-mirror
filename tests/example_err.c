/*
 * Test err(), verr(), ... Non-standard, BSD only.
 *
 * Copyright (C) 2013-2018  Simon Ruderich
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include "../src/compiler.h"
#include "example.h"


int main(int argc unused, char **argv unused) {
    pid_t pid;

    FORKED_TEST(pid) { errno = ENOMEM; err(0, "error: %s", "message"); }
    FORKED_TEST(pid) { errno = ENOMEM; err(1, "error: %s", "message"); }

    FORKED_TEST(pid) { errx(0, "error: %s", "message"); }
    FORKED_TEST(pid) { errx(1, "error: %s", "message"); }

    errno = ENOMEM;
    warn("warning: %s", "message");
    warnx("warning: %s", "message");

    /* v*() functions are implicitly tested - the implementation uses them. */

    printf("\n");
    return EXIT_SUCCESS;
}
