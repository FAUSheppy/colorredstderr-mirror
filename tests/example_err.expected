>STDERR>example_err: error: message: Cannot allocate memory
<STDERR<exit code: 0
>STDERR>example_err: error: message: Cannot allocate memory
<STDERR<exit code: 1
>STDERR>example_err: error: message
<STDERR<exit code: 0
>STDERR>example_err: error: message
<STDERR<exit code: 1
>STDERR>example_err: warning: message: Cannot allocate memory
example_err: warning: message
<STDERR<
